File: pyround1/1005.k-次取反后最大化的数组和.py
#
# @lc app=leetcode.cn id=1005 lang=python3
#
# [1005] K 次取反后最大化的数组和
#

# @lc code=start
class Solution:
    def largestSumAfterKNegations(self, A: List[int], K: int) -> int:
        A.sort()
        i = 0
        while i < K:
            if A[i] < 0:
                A[i] = -A[i]
                i += 1
            else:
                if (K - i) %2 == 0:
                    break
                else:
                    if i != 0 and A[i] > A[i - 1]:
                        A[i-1] = -A[i-1]
                    else:
                        A[i] = -A[i]
                    break
        return sum(A)
        
# @lc code=end

File: 1005.k-次取反后最大化的数组和.cpp
/*
 * @lc app=leetcode.cn id=1005 lang=cpp
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start
class Solution {
public:
    int largestSumAfterKNegations(vector<int>& A, int K) {
        sort(A.begin(), A.end());
        int pos = 0;
        while(K){
            if(A[pos] < 0){
                if(A[pos+1] > 0)
                {
                    if(abs(A[pos]) < A[pos+1])
                    {
                        if(K%2 == 1)
                        {
                            A[pos] = -A[pos];
                        }
                        break;
                    }
                }
                A[pos] = -A[pos];
                K--;
                pos++;
            }
            else{
                if(K % 2 == 1)
                    A[pos] = -A[pos];
                break;
            }
        }
        int sum = 0;
        for(int num : A){
            cout << num << " ";
            sum += num;
        }
        return sum;
    }
};
// @lc code=end

